---
title: "phyloseq"
author: "Vanessa Dumeaux"
date: "6/15/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(rprojroot)
root<-rprojroot::find_root(".git/index")

setwd(file.path(root,"exp/exp05-phyloseq"))   


suppressPackageStartupMessages({
  library(curatedMetagenomicData)
  library(phyloseq)
  library(dplyr)
  library(ggplot2)
  library(ape)
  library(TreeSummarizedExperiment)
  library(dplyr)
  library(treeclimbR)
  library(ggtree)
  library(TreeHeatmap)
  library(ggplot2)
  library(viridis)
  library(ggnewscale)
})


```

# Download data

```{r}
datdir <- "/home/data/refined/microbiome2brain/microbiome/01-trimmed/metaphlan"

dat <- read.csv(file.path(datdir, "merged_abundance_table.txt"), sep = "\t", skip = 1, header = TRUE, row.names = 1)
dat <- dat[,-1]
colnames(dat) <- gsub(".profiled_metagenome", "", colnames(dat))

sn <- colnames(dat)
sn <- unlist(lapply(strsplit(sn, "_"), "[", 1))
participant <- substr(sn,1,2)
timepoint <- substr(sn,3,3)

nreads <- read.csv("/home/data/refined/microbiome2brain/microbiome/01-trimmed/nreads.csv", header=TRUE, stringsAsFactors=FALSE)
nreads$sampleID <- unlist(lapply(strsplit(nreads$sampleID, "_"), "[", 1)) 

pdata <- data.frame(sampleID=sn,
                    personID=participant,
                    timepoint=timepoint, stringsAsFactors = FALSE)

pdata <- left_join(pdata, nreads)
rownames(pdata) <- colnames(dat)

## other metadata
final <- read.csv("/home/data/refined/microbiome2brain/metadata/final.txt", sep="\t")
final$sampleID <- final$microbiomeID
dexa <- read.csv("/home/data/refined/microbiome2brain/metadata/90000152 DEXA DATA.txt", sep="\t")
dexa$sampleID <- gsub("_", "", dexa$sampleID)
dexa$sampleID <- gsub("N15200|R15200", "", dexa$sampleID)
all <- left_join(final, dexa)

pdata0 <- pdata

pdata <- left_join(pdata, all)
rownames(pdata) <- colnames(dat)


stool_0 <- ExpressionSet(assayData=as.matrix(dat), phenoData=new("AnnotatedDataFrame",pdata))
stool <- ExpressionSet2phyloseq(stool_0, phylogenetictree = TRUE)
saveRDS(stool,"../../data/stool.rds")
```

```{r}

p0 <- plot_bar(stool, fill = "Family") +
theme_minimal()
getwd()

ggsave("output/RA_barplot.pdf", p0)


p1 <- plot_heatmap(stool, "NMDS", "bray", sample.label = "sampleID", taxa.label = "Family",low="#66CCFF", high="#000033", na.value="white", first.sample = "E2A_A3")

ggsave("output/RA_heatmap_family.pdf", p1)


sample_data(stool)$num.participant <- as.numeric(sample_data(stool)$participant)

set.seed(711L)
p2 <- plot_net(stool, point_label = "sampleID", color = "personID", shape = "timepoint", maxdist=0.4)

ggsave("output/RA_network.pdf", p2)


```
```{r}
ord <- UniFrac(stool)
stool.pcoa = ordinate(stool, method="PCoA", distance=ord)
plot_scree(stool.pcoa, "Scree plot for stool, UniFrac/PCoA")

p12 <- plot_ordination(stool, stool.pcoa, type="samples", color = "personID") + 
  theme_minimal() +  geom_text(aes(label = sampleID), position = position_nudge(y = -0.005))+
  geom_point(size=2) + geom_path() + scale_colour_hue(guide = FALSE)

p13 <- plot_ordination(stool, stool.pcoa, type="samples", color = "personID", axes=c(1, 3)) + 
   theme_minimal() +  geom_text(aes(label = sampleID), position = position_nudge(y = -0.005))+
  geom_point(size=2) + geom_path() + scale_colour_hue(guide = FALSE) + theme_minimal()

p23 <- plot_ordination(stool, stool.pcoa, type="samples", color = "personID", axes=c(2, 3)) + 
   theme_minimal() +  geom_text(aes(label = sampleID), position = position_nudge(y = -0.005))+
  geom_point(size=2) + geom_path() + scale_colour_hue(guide = FALSE) + theme_minimal()

```

# Data preparation
# raw counts
```{r}
otu_count <- as.matrix(data.frame(otu_table(stool)))
reads <- pdata$nreads
new_count <- round((otu_count/100) %*% diag(reads), digits = 0)
colnames(new_count) <- colnames(otu_count)
```

```{r}
# taxonomic table
new_tax <- data.frame(tax_table(stool)@.Data, check.names = FALSE)

# phylogenetic tree
new_phy <- phy_tree(stool)

```



Data is stored as a `TreeSummarizedExperiment` object

```{r}
lse_phy <- TreeSummarizedExperiment(assays = list(count = new_count),
                                    rowData = new_tax,
                                    colData = pdata,
                                    rowTree = new_phy,
                                    rowNodeLab = rownames(new_count))
```

Data on internal nodes are generated by data on leaves.
```{r}
all_node <- showNode(tree = new_phy, only.leaf = FALSE)
tse <- aggValue(x = lse_phy, rowLevel = all_node)
```

# run `edgeR` 

For data in each category, we compare the microbial abundance between two different delivery methods (vaginal and c-section) using `edgeR`.

```{r}

# build model
res <- runDA(TSE = tse, feature_on_row = TRUE, assay = 1, design_terms = c("participant", "timepoint"))

# extract the result table
out <- nodeResult(res, n = Inf)


```

# run `treeclimbR`

There are two steps: propose candidates and evaluate candidates.
```{r}
# get candidates
cand <- getCand(tree = new_phy, score_data = out, 
                node_column = "node", 
                p_column = "PValue", 
                sign_column = "logFC")


# evaluate candidates

best <- evalCand(tree = new_phy, type = "single",
                 levels = cand$candidate_list,
                 limit_rej = 0.05,
                 score_data = out,
                 node_column = "node",
                 p_column = "PValue",
                 sign_column = "logFC",
                 use_pseudo_leaf = FALSE)

names(best) <- c("condidate_list", "output")

```

# Results 


```{r}
# results of treeclimbR
df <- topNodes(object = best, n = nrow(res), p_value = 0.05)
lapply(df, head)

# nodes identified
loc <- df$node
lapply(loc, length)

# the number of descendant leaves of identified nodes
xx <- findOS(tree = new_phy, node = loc, only.leaf = TRUE, self.include = TRUE)
               ux <- unlist(xx)
               length(unique(ux))
```


Here outputs results of treeclimbR for visualization in next step
```{r}
save(bse, loc, file = "output/Analysis.RData")
```


# Plot tree
```{r}
new_phy <- rowTree(tse)

# tree
fig_0 <- ggtree(new_phy, layout = "fan", size = 0.5,
                 branch.length = "none", 
                 open.angle = 50) %>% rotate_tree(140)

fig_0

```

# Color branches and nodes

Branches that are detected in at least one time point are colored. The detected
nodes in period `0 M` are labeled as red points.
```{r}
# tree: color branches that are detected in at least one period
uloc <- unique(unlist(loc))
ubr <- findOS(tree = new_phy, node = uloc, 
              only.leaf = FALSE, self.include = TRUE)
ubr <- unlist(ubr)

all_node <- showNode(tree = new_phy, only.leaf = FALSE)
df_br <- data.frame(node = all_node, Detected = "F") %>%
  mutate(Detected = ifelse(node %in% ubr, "Yes", "No"))


# tree (colored branch) + result
fig_0 <- fig_0 %<+% df_br + aes(color = Detected)+
    geom_point2(aes(subset = (node %in% loc[[1]])), 
                      color = "red", size = 2) +
  scale_color_manual(values = c("Yes" = "orange", "No" = "grey")) +
  guides(color = guide_legend(override.aes = list(size = 2)),
         fill = guide_legend(order = 1))


fig_0
```


# Black bars

Below add bars to indicate branches detected in different periods.

```{r}
# the alias labels of leaves on detected branches.

lx <- findOS(tree = new_phy, node = loc, 
               only.leaf = TRUE, self.include = TRUE)
lx <- unlist(lx)
loc_leaf_alias <- transNode(tree = new_phy, node = lx, use.alias = TRUE)


# A binary matrix: Yes (DA) & No (non-DA)
leaf <- showNode(tree = new_phy, only.leaf = TRUE)
leaf_alias <- transNode(tree = new_phy, node = leaf, use.alias = TRUE)
mat_result <- matrix(NA, nrow = length(leaf), ncol = 1)
rownames(mat_result) <- leaf_alias
colnames(mat_result) <- "stool"

for (i in seq_along(loc_leaf_alias)) {
  mat_result[, i] <- ifelse(leaf_alias %in% loc_leaf_alias[[i]], "Yes", "No")
}

head(mat_result)
```

```{r}
# tree + bar (identified by treeclimbR)
col_sp <- colnames(mat_result)
names(col_sp) <- col_sp
fig_1 <- TreeHeatmap(tree = new_phy, 
                     tree_fig = fig_0, 
                     hm_data = mat_result, 
                     rel_width = 0.2,
                     tree_hm_gap = 0.3,
                     column_split = col_sp,
                     column_split_gap = 0.1,
                     show_colnames = TRUE,
                     colnames_position = "top",
                     colnames_size = 2, 
                     colnames_angle = 10, 
                     colnames_hjust = 0.5,
                     colnames_offset_y = 20,
                     colnames_offset_x = 0.25) +
  scale_fill_manual(values = c("Yes" = "black", 
                               "NO" = "white"), guide = FALSE) +
  new_scale_fill()

fig_1
```



# Add heatmaps

Average samples that are generated by randomly assigning samples that have the born method and at the same time point to five categories for each category. 

```{r}
# observed data: count (in cpm)
lse <- tse[rowLinks(tse)$isLeaf, ]

count <- edgeR::cpm(assays(lse)[[1]])


# Average samples: samples that have the born method and at the same time point are randomly assigned to five categories and an average sample is generated for each category.  
set.seed(1)

# scale count: log transform + scale

xx <- count
lx <- log10(xx + 1) 
ax <- apply(lx, 1, FUN = function(y) {
  #if (length(unique(y))) {y} else {scale(y)}
    ly <- length(unique(y))
    if (ly == 1) {
      ifelse(y > 0 , 1, 0)
    } else{
      sy <- scale(y)
      (sy - min(sy))/(max(sy) - min(sy))
      }
    })

tx <- t(ax)
colnames(tx) <- colnames(xx)

```


```{r}
fig_2 <- fig_1
angle <- c(8, 5, 5)
  
split_i <- pdata$timepoint
names(split_i) <- colnames(tx)
  
split_lab_i <- split_i
names(split_lab_i) <- split_i
  
  # 6 20
fig_2 <- TreeHeatmap(tree = new_phy, 
              tree_fig = fig_2,  
              tree_hm_gap = 2,
              column_split_gap = 0.2, 
              column_split = split_i,
              column_split_label = split_lab_i,
              split_label_size = 3,
              split_label_offset_y = 22 -(i) * 3 ,
              split_label_offset_x = -1,
              split_label_angle = angle[i],
              rel_width = 0.5,
              hm_data = tx,
              legend_title_hm = "scaled log-CPM") +
    guides(fill = guide_legend(order = 2))
  

fig_2 

ggsave("output/treeclimbr-timepoint.pdf", fig_2)
```
# Anotate genus

The genus of detected branches. We only show genus that has above five species (leaf nodes) detected.
```{r, warning=FALSE, message=FALSE}
# genus: detected

des.i <- findOS(tree = new_phy, node = loc, only.leaf = TRUE, self.include = TRUE, use.alias = TRUE)

ri <- rowLinks(tse)$nodeNum %in% des.i
  
bse.ri <- tse[ri, ]
df_genus <- data.frame(Genus = as.character(rowData(bse.ri)$Genus),
                       node = rowLinks(bse.ri)$nodeNum)
  

# the matrix is used later to label heatmap
mat_genus <- cbind(as.character(df_genus$Genus))
colnames(mat_genus) <- "Genus"
rownames(mat_genus) <- df_genus$alias

```
